core/src/main/java/io/grpc/transport/AbstractStream.java

```diff
@@ -351,28 +345,27 @@ protected final void onSendingBytes(int numBytes) {
   protected final void onSentBytes(int numBytes) {
     boolean doNotify;
     synchronized (onReadyLock) {
+      doNotify = wentBelowThreshold(
+          numSentBytesQueued,  // Previous queued sent bytes
+          onReadyThreshold,
+          numSentBytesQueued - numBytes, // New queued sent bytes
+          onReadyThreshold);
       numSentBytesQueued -= numBytes;
-      doNotify = needToNotifyOnReady();
     }
     if (doNotify) {
-      listener().onReady();
+      maybeNotifyListener();
```

Comment:
~technically~ We don't.  It would be possible that something could acquire the onReadyLock after the synchronized block (a onSendBytes, perhaps), in which case it might no longer be appropriate to notify.  Also, I believe onReady was supposed to be called outside the context of the lock, but I may have imagined that.